lib/application.js

```diff
@@ -567,3 +571,65 @@ app.listen = function(cb) {
 
   return server;
 };
+
+/* START CONTROLLER CODE*/
+
+// this is similar to app.model
+// at the moment I haven't really used 'options'
+app.controller = function(ctor, options) {
+  var self = this;
+
+  // the SharedController is a wrapper, see below
+  var controller = new SharedController(ctor);
+  this.remotes().addClass(controller.sharedClass);
+  app.controllers().push(controller);
+
+  return controller;
+
+  // SharedController class contains strong-remoting objects to set up the metadata
+  // and register the endpoint, etc
+  function SharedController(ctor, options) {
+    options = options || {};
+    this.name = ctor.name;
+    this.ctor = ctor;
+    // TODO - right now there's a hack on line 289 to skip this
+    // util.inherits(ctor, EventEmitter);
+
+    this.sharedClass = new RemoteObjects.SharedClass(this.name, ctor, options);
+    // the resolveScope method instantiates the controller upon request
+    // (please correct my terminology if it's wrong)
+    this.sharedClass.resolveScope = function(remotingContext) {
+      // self is the app - should it be something more obvious like appInstance?
+      return new ctor(self, remotingContext);
+    };
+
+    // similar to model.remoteMethod
+    this.remoteMethod = function(name, options) {
+      // I didn't dive into why, but this only worked when I set isStatic to true
+      // whether the controller methods were prototype methods or not
+      options.isStatic = true;
+      this.sharedClass.defineMethod(name, options);
+    };
+
+    // have not tested this method yet but it follows model.disableRemoteMethod
+    this.disableRemoteMethod = function(name, isStatic) {
+      this.sharedClass.disableMethod(name, isStatic || false);
+      // this.emit('remoteMethodDisabled', this.sharedClass, name);
+    };
+  };
+};
+
+// similar to this app.models
+// the controllers still need to registered in the Registy though
+app.controllers = function() {
```

Comment:
`app.models` has dual API: 

1) it's a function you can call to receive an array of models:

```js
app.models().forEach(function(model) {
  // do something with a model
});
```

2) And at the same time, it's a name-based lookup table allowing quick access to any model:

```js
var MyUser = app.models['my-user'];
// app.models.myUser and app.models.MyUser point to the same model ctor
```

This duality is causing a bit of confusion, one of our colleagues was surprised to learn that `Object.keys(app.models)` contains each model two or three times.

I think it would be great if we can come up with a better API for controllers.

Questions to consider:
 - Do we need a name-based lookup for controllers?
 - If the answer is yes, do we need API for getting a list of controllers, or is it enough to build this list via `Object.keys(app.controllers)`?
 - Can we require controller names to be unique? Name-based lookup won't work otherwise.
 - I am proposing to remove classify/camelize aliases and always use the original controller name.
 - Since it's not possible to change function name after it was defined, we need to use a different property name to allow code-generated controller functions to provide custom name. For example, models use `modelName` property.
 
@richardpringle @ritch @raymondfeng thoughts?